File: Course5_BackEnd(kotlin_spring)/ch01_kotlin/java-example/src/main/java/org/example/ex02/Exam02.java
package org.example.ex02;

import java.util.Optional;

public class Exam02 {
    private int a; // 0

    Exam02() {
        // 코드 작성
        var b = 20; // 타입 추론
//        var c = null; // X
        int c = 30;
//        int d;
        Integer e = new Integer(100);
        Integer f = 20;
        Integer g = null;

        callFunction(a);
        callFunction(b);
        callFunction(c);
//        callFunction(d);
        callFunction(e);
        callFunction(f);
        callFunction(g); //NPE
        callFunction(null);

    }

    public void callFunction(Integer i) {
        var _i = (i == null) ? Integer.valueOf(100) : Integer.valueOf(i);
        var temp = _i;
        System.out.println(temp * 20);
    }
    public static void main(String[] args) {
        new Exam02();
    }
}

File: Course5_BackEnd/ch01_kotlin/kotlin-example/bin/main/ex02/Exam02.kt
package ex02

import java.util.Optional

fun main() {
    val a: Int = 0
    val b = 10
    val c: Int = 20
    val d: Int? = null

    callFunction(a)
    callFunction(b)
    callFunction(c)
    callFunction(d)
    callFunction()

}

// '?' 엘비스 연산자(Elvis Operation)
fun callFunction(i: Int? = 100) {
    // ? << null 이 올수도 있다
    // 변수? << 변수가 null 이야?
    // 변수?. << 변수가 null 이 아닐때
    // 변수?: << 변수가 null 일때

//    val temp = if (i == null) "null 값 입니다" else i
    val temp = i?.let { it * 20 } ?: "null 값 입니다"
    println(temp)

    /*
    i?.let {
        println(it)
    }?: run{
        println(null)
    }
    println(i)
    */
}
